arroyo-connectors/src/kafka.rs

```diff
@@ -0,0 +1,399 @@
+        let client = self.connect().await?;
+
+        self.info("Connected to Kafka").await;
+
+        let topic = self.table.topic.clone();
+
+        let metadata = client
+            .fetch_metadata(Some(&topic), Duration::from_secs(10))
+            .map_err(|e| format!("Failed to fetch metadata: {:?}", e))?;
+
+        self.info("Fetched topic metadata").await;
+
+        {
+            let topic_metadata = metadata.topics().get(0).ok_or_else(|| {
+                format!(
+                    "Returned metadata was empty; unable to subscribe to topic '{}'",
+                    topic
+                )
+            })?;
+
+            if let Some(err) = topic_metadata.error() {
+                match err {
+                    rdkafka::types::RDKafkaRespErr::RD_KAFKA_RESP_ERR__UNKNOWN_PARTITION
+                    | rdkafka::types::RDKafkaRespErr::RD_KAFKA_RESP_ERR__UNKNOWN_TOPIC
+                    | rdkafka::types::RDKafkaRespErr::RD_KAFKA_RESP_ERR_UNKNOWN_TOPIC_OR_PART => {
+                        return Err(format!(
+                            "Topic '{}' does not exist in the configured Kafka cluster",
+                            topic
+                        ));
+                    }
+                    e => {
+                        error!("Unhandled Kafka error while fetching metadata: {:?}", e);
+                        return Err(format!(
+                            "Something went wrong while fetching topic metadata: {:?}",
+                            e
+                        ));
+                    }
+                }
+            }
+
+            let map = topic_metadata
+                .partitions()
+                .iter()
+                .map(|p| ((topic.clone(), p.id()), Offset::Beginning))
+                .collect();
+
+            client
+                .assign(&TopicPartitionList::from_topic_map(&map).unwrap())
+                .map_err(|e| format!("Failed to subscribe to topic '{}': {:?}", topic, e))?;
+        }
+
+        if let TableType::Source { .. } = self.table.type_ {
+            self.info("Waiting for messsages").await;
+
+            let start = Instant::now();
+            let timeout = Duration::from_secs(30);
+            while start.elapsed() < timeout {
+                match client.poll(Duration::ZERO) {
+                    Some(Ok(message)) => {
+                        self.info("Received message from Kafka").await;
+                        self.test_schema(message)?;
+                        return Ok(());
+                    }
+                    Some(Err(e)) => {
+                        warn!("Error while reading from kafka in test: {:?}", e);
+                        return Err(format!("Error while reading messages from Kafka: {}", e));
+                    }
+                    None => {
+                        // wait
+                    }
+                }
+
+                tokio::time::sleep(Duration::from_millis(500)).await;
```

Comment:
Could just move this into the None block.

Reply:
sure